docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,197 @@
+      # https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#an-example-of-a-pod-that-uses-pod-affinity
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+            - key: "foundationdb.org/distribution-key"
+              operator: NotIn
+              values:
+              - "all-0"
+            - key: "foundationdb.org/fdb-cluster-name"
+              operator: In
+              values:
+              - "example-cluster"
+          topologyKey: fault_domain
+        weight: 1
+```
+
+The additional label and the `affinity` term enable the operator to spread Pods across physical fault domains.
+The following code snippet shows a possible implementation of the required structs:
+
+```golang
+type BinPack string
+
+const (
+    // BinPackAll defines that processes independent of their class should be bin packed.
+    BinPackAll BinPack = "All"
+    // BinPackStateful defines that stateful processes should be bin packed and additionally stateless processes.
+    // The prefix will be either "stateful" or "stateless".
+    BinPackStateful BinPack = "Stateful"
+    // BinPackClass only processes of the same class will be bin packed.
+    // As prefix we will use the according process class.
+    BinPackClass BinPack = "Class"
+)
+
+// DistributionConfig
+type DistributionConfig struct {
+    // Enabled defines if the binpacking is enabled or not.
+    // Default: false
+    Enabled *bool
+    // DesiredFaultDomains desfines the number of desired fault domain.
+    // Must be greater than 0 if fault domain distribution is enabled.
+    // Default: 1
+    DesiredFaultDomains *int
+    // BinPack defines what processes should be bin packed to the fault domains.
+    // Default: "Class"
+    BinPack *BinPack
+    // Required defines if the affinity terms are required or preffered.
+    // Default: false
+    Required *bool
+}
+```
+
+The operator will try to spread the Pods equally across the logical fault domains.
+Depending on the replacements for unreachable Pods we could use one logical fault domain more than another.
+We also don't guarantee that we use exactly the number of `DesiredFaultDomains` e.g. when we spawn less processes than `DesiredFaultDomains`.
+
+A change to `DesiredFaultDomains` will lead to a migration of a subset of Pods in order to honor the new affinity term.
+The operator only tries to replace as many Pods as required to have all pods equally distributed across the logical fault domains.
+
+## Scheduling and Eviction Policy
+
+For the implementation we need two additional steps:
+
+- The `AddProcessGroups` reconcile loop must be aware of the fault domains.
+- The `ReplaceMisconfiguredProcessGroups` reconcile loop must be aware of the fault domains.
+
+We will only count process groups that are not marked for removal.
+We can calculate the minimum and maximum number of Pods per fault domain with the following logic:
```

Comment:
Is this calculated separately for every process class? I'm a little confused about how that interacts with the different bin packing strategies.